File: Word-Formed/src/com/google/corrigan/owen/wordformed/DraggableBox.java
package com.google.corrigan.owen.wordformed;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.media.AudioManager;
import android.media.SoundPool;
import android.media.SoundPool.OnLoadCompleteListener;
import android.os.SystemClock;
import android.os.Vibrator;
import android.view.MotionEvent;

public class DraggableBox
{
	//Top left X and Y coordinates of draggable box
	
	private static SoundPool sound;
	static int popID;
	static int submitID;
	static int enteredID;
	static int finishID;
	private static boolean loaded;

	
	private float rectX;
	float rectY;
	//Hold starting position when the box is picked up
	private float startX;
	private float startY;
	//Length of rectangle
	private int rectSize = 45;
	//Size of border
	private int borderSize = 5;
	//Outer and inner rectangles used for dragging
	private RectF rect;
	//Boolean to say if box is being dragged
	private boolean dragging = false;
	//Letter stored in tile
	private char letter;
	private static Context context;
	private Bitmap tile;
	private Paint paint;
	private float speed = 5;
	private float velX;
	private float velY;
	//True when user lets go
	private boolean flinging = false;
	private int transperancy = 255;
	private boolean movingLeft = false;
	private boolean movingRight = false;
	private float targetX = 0;
	private int value;
	
	boolean notMoved = true;
	//Constructor. Takes starting position as parameters
	public DraggableBox(Context context0, float topX, float topY)
	{
		//Set coordinates of box
		rectX = topX;
		rectY = topY;
		//Set coordinates in case box needs to jump back to this position
		startX = topX;
		startY = topY;
		//Calculate size and position of outer and inner box
		rect = new RectF(rectX, rectY, rectX+rectSize, rectY+rectSize);
		//Randomly choose a character
		letter = TileGenerator.nextTile();
		value = TileGenerator.getValue(letter);
		context = context0;	
		tile = BitmapFactory.decodeResource(context.getResources(), R.drawable.tile);
		paint = new Paint();
		paint.setColor(Color.BLACK);
		paint.setTextSize(30);
		paint.setTypeface(Typeface.MONOSPACE);
		
		sound = new SoundPool(10, AudioManager.STREAM_MUSIC, 0);
		sound.setOnLoadCompleteListener(new OnLoadCompleteListener() 
		{
			public void onLoadComplete(SoundPool soundPool, int sampleId, int status) 
			{
				loaded = true;
			}
		});
		popID = sound.load(context, R.raw.pop_sound, 1);
		submitID = sound.load(context, R.raw.submit, 1);
		enteredID = sound.load(context, R.raw.entered, 1);
		finishID = sound.load(context, R.raw.finish, 1);
	}

	
	public char getLetter()
	{
		return letter;
	}
	
	//Draw method. Takes canvas as parameter
	public synchronized void draw(Canvas canvas)
	{
		//Draw Outer Rectangle
		//Draw inner Rectangle
		Drawable box = context.getResources().getDrawable(R.drawable.tile);
		tile = ((BitmapDrawable)box).getBitmap();
		paint.setAlpha(transperancy);
		canvas.drawBitmap(tile, null, rect, paint);
		//Draw letter
		if(!dragging && !flinging)
			canvas.drawText(letter+"", rectX + 15, rectY + 35, paint);
		else
			canvas.drawText(letter+"", rectX + 15 - 35, rectY + 35 - 10, paint); //COME BACK TO THIS
	}
	
	//On touch event. Takes MotionEvent and Dropbox as parameter
	public synchronized boolean onTouchEvent(MotionEvent event, Dropbox drop, Dropbox answer, CreateBox create)
	{
		//Store mouse x and y coordinates
		float mouseX = event.getX();
		float mouseY = event.getY();
		switch(event.getAction())
		{
			//On mouse down, check if touched within area. If so start dragging
			case MotionEvent.ACTION_DOWN:
				startX = mouseX;
				startY = mouseY;
				if(rect.contains(mouseX, mouseY) && !flinging)
				{
					playSound(popID);	
					Vibrator v = (Vibrator) context.getSystemService(Context.VIBRATOR_SERVICE);
					v.vibrate(60);
					dragging = true;
					rectSize = 80;
					paint.setTextSize(60);
					//Update position immediately to prevent letter inside from jumping
					rectX = mouseX;
					rectY = mouseY;
					rect = new RectF(rectX - 40, rectY - 40, rectX+rectSize - 40, rectY+rectSize - 40);
					new RectF(rectX + borderSize - 40, rectY + borderSize - 40, 
							rectX + rectSize - borderSize - 40, rectY + rectSize - borderSize - 40);
				}
				else dragging = false;
				break;
			/* On mouse up drop tile in place if in valid location.
			 * If it landed on another tile swap that tile back to it's place
			 * Or be deleted otherwise
			*/
			case MotionEvent.ACTION_UP:
				//angle = Math.atan(mouseY - startY  / mouseX - startX);
				if(dragging)
				{
					velY = (mouseY - startY);
					velX = (mouseX - startX);
					float factor = (float)Math.sqrt(velX * velX + velY * velY) / speed;
					velX /= factor;
					velY /= factor;
					
					drop.remove(this);
					answer.remove(this);
					if(create.contains(this))
						create.remove(this);
					
					//if within drop zone snap to grid
					if(drop.contains(rectX, rectY))
					{
						playSound(popID);
						drop.add(this);
						notMoved = false;
						rectSize = 45;
						paint.setTextSize(30);
						rect = new RectF(rectX, rectY, rectX+rectSize, rectY+rectSize);
						new RectF(rectX + borderSize, rectY + borderSize, 
								rectX + rectSize - borderSize, rectY + rectSize - borderSize);
					}
					else if(answer.contains(rectX, rectY))
					{
						playSound(popID);
						answer.add(this);
						rectSize = 45;
						paint.setTextSize(30);
						rect = new RectF(rectX, rectY, rectX+rectSize, rectY+rectSize);
						new RectF(rectX + borderSize, rectY + borderSize, 
								rectX + rectSize - borderSize, rectY + rectSize - borderSize);
						
					}
					//If box not within dropbox, remove from board
					//TODO: return null reference to prevent memory leak
					else
					{
						flinging = true;
					}
				}
				//Fixes bug for some reason
				//TODO: figure out what next line does
				dragging = false;
				break;
			//If this is being dragged, set position to same position as mouse
			//TODO: Instead of default, find specific dragging case
			default:
				if(dragging)
				{
					rectX = mouseX;
					rectY = mouseY;
					rect = new RectF(rectX - 40, rectY - 40, rectX+rectSize - 40, rectY+rectSize - 40);
					new RectF(rectX + borderSize - 40, rectY + borderSize - 40, 
							rectX + rectSize - borderSize - 40, rectY + rectSize - borderSize - 40);
				}
		}
		return true;
	}
	
	public void anim()
	{
		if(flinging)
		{
			rectX += velX;
			rectY += velY;
			rect = new RectF(rectX - 40, rectY - 40, rectX+rectSize - 40, rectY+rectSize - 40);
			new RectF(rectX + borderSize - 40, rectY + borderSize - 40, 
					rectX + rectSize - borderSize - 40, rectY + rectSize - borderSize - 40);
			if(transperancy > 0)
				transperancy -= 5;
		}
		else if(movingLeft)
		{
			if(rectX <= targetX)
			{
				rectX = targetX;
				movingLeft = false;
				velX = 0;
			}
			rectX += velX;
			rect = new RectF(rectX, rectY, rectX + rectSize, rectY + rectSize);
			new RectF(rectX + borderSize, rectY + borderSize, 
					rectX + rectSize - borderSize, rectY + rectSize - borderSize);
		}
		else if(movingRight)
		{
			if(rectX >= targetX)
			{
				rectX = targetX;
				movingRight = false;
				velX = 0;
			}
			rectX += velX;
			rect = new RectF(rectX, rectY, rectX + rectSize, rectY + rectSize);
			new RectF(rectX + borderSize, rectY + borderSize, 
					rectX + rectSize - borderSize, rectY + rectSize - borderSize);
		}
	}
	
	public int getValue()
	{
		return value;
	}
	
	public boolean isSelected()
	{
		return dragging;
	}
	
	//Move this tile to a specific x and y position. Takes x and y position as parameters
	private void moveLeft(float x, float y)
	{
		targetX = x;
		movingLeft = true;
		
		velX = -1 * speed;
		//rectX = x;
		rectY = y;
		rect = new RectF(rectX, rectY, rectX + rectSize, rectY + rectSize);
		new RectF(rectX + borderSize, rectY + borderSize, 
			rectX + rectSize - borderSize, rectY + rectSize - borderSize);
	}
	
	private void moveRight(float x, float y)
	{
		targetX = x;
		movingRight = true;
		
		velX = speed;
		//rectX = x;
		rectY = y;
		rect = new RectF(rectX, rectY, rectX + rectSize, rectY + rectSize);
		new RectF(rectX + borderSize, rectY + borderSize, 
			rectX + rectSize - borderSize, rectY + rectSize - borderSize);
	}
	
	public void move(float x, float y)
	{
		if(rectX > x)
			moveLeft(x, y);
		else
			moveRight(x, y);
	}
	
	public static void playSound(int soundID)
	{
		if(loaded)
		{
			AudioManager am = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
			float curVol = (float) am.getStreamVolume(AudioManager.STREAM_MUSIC);
			float maxVol = (float) am.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
			float vol = curVol/maxVol;
			int waitCounter = 0;
			int waitLimit = 20;
			while(sound.play(soundID, vol, vol, 1, 0, 1.f) == 0 && waitCounter < waitLimit)
			{
				waitCounter++;
				SystemClock.sleep(5);
			}
		}				
	}
}

File: app/src/main/java/com/google/corrigan/owen/wordformed/DraggableBox.kt
package com.google.corrigan.owen.wordformed

import android.content.Context
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.SoundPool
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.view.MotionEvent

class DraggableBox(private val context: Context, topX: Float, topY: Float) {
    private var rectX: Float = topX
    private var rectY: Float = topY
    private var startX: Float = topX
    private var startY: Float = topY
    private var rectSize = 45
    private val borderSize = 5
    private var rect: RectF = RectF(rectX, rectY, rectX + rectSize, rectY + rectSize)
    private var dragging = false
    private val letter: Char = TileGenerator.nextTile()
    private val value: Int = TileGenerator.getValue(letter)
    private var tile: Bitmap? = null
    private val paint: Paint = Paint()
    private val speed = 5f
    private var velX = 0f
    private var velY = 0f
    private var color: Int = 0
    private var flinging = false
    private var transparency = 255
    private var movingLeft = false
    private var movingRight = false
    private var targetX = 0f
    var notMoved = true

    init {
        val box = context.resources.getDrawable(R.drawable.tile, null)
        tile = (box as BitmapDrawable).bitmap
        
        color = when {
            value == 1 -> Color.BLACK
            value == 2 -> Color.rgb(60, 0, 0)
            value == 3 -> Color.rgb(120, 0, 0)
            value == 4 -> Color.rgb(180, 0, 0)
            value > 4 -> Color.rgb(255, 0, 0)
            else -> Color.BLACK
        }

        paint.color = color
        paint.textSize = 30f
        paint.typeface = Typeface.MONOSPACE
    }

    companion object {
        private var loaded = false
        private var soundPool: SoundPool? = null
        var popID: Int = 0
        var submitID: Int = 0
        var enteredID: Int = 0
        var finishID: Int = 0

        fun initSound(context: Context) {
            if (soundPool != null) return

            val audioAttributes = AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_GAME)
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .build()

            soundPool = SoundPool.Builder()
                .setMaxStreams(10)
                .setAudioAttributes(audioAttributes)
                .build()

            soundPool?.setOnLoadCompleteListener { _, _, _ -> loaded = true }

            popID = soundPool?.load(context, R.raw.pop_sound, 1) ?: 0
            submitID = soundPool?.load(context, R.raw.submit, 1) ?: 0
            enteredID = soundPool?.load(context, R.raw.entered, 1) ?: 0
            finishID = soundPool?.load(context, R.raw.finish, 1) ?: 0
        }

        fun playSound(context: Context, soundID: Int) {
            if (loaded) {
                val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
                val curVol = am.getStreamVolume(AudioManager.STREAM_MUSIC).toFloat()
                val maxVol = am.getStreamMaxVolume(AudioManager.STREAM_MUSIC).toFloat()
                val vol = curVol / maxVol
                soundPool?.play(soundID, vol, vol, 1, 0, 1f)
            }
        }
    }

    fun getLetter(): Char = letter

    @Synchronized
    fun draw(canvas: Canvas) {
        paint.alpha = transparency
        tile?.let {
            canvas.drawBitmap(it, null, rect, paint)
        }
        
        if (!dragging && !flinging) {
            canvas.drawText(letter.toString(), rectX + 15, rectY + 35, paint)
        } else {
            canvas.drawText(letter.toString(), rectX + 15 - 35, rectY + 35 - 10, paint)
        }
    }

    @Synchronized
    fun onTouchEvent(event: MotionEvent, drop: Dropbox, answer: Dropbox, create: CreateBox): Boolean {
        val mouseX = event.x
        val mouseY = event.y
        
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                startX = mouseX
                startY = mouseY
                if (rect.contains(mouseX, mouseY) && !flinging) {
                    vibrate(60)
                    playSound(context, popID)
                    dragging = true
                    rectSize = 80
                    paint.textSize = 60f
                    rectX = mouseX
                    rectY = mouseY
                    updateRect(true)
                } else {
                    dragging = false
                }
            }
            MotionEvent.ACTION_UP -> {
                if (dragging) {
                    velY = mouseY - startY
                    velX = mouseX - startX
                    val factor = Math.sqrt((velX * velX + velY * velY).toDouble()).toFloat() / speed
                    velX /= factor
                    velY /= factor

                    drop.remove(this)
                    answer.remove(this)
                    if (create.contains(this)) {
                        create.remove(this)
                    }

                    if (drop.contains(rectX, rectY)) {
                        playSound(context, popID)
                        drop.add(this)
                        notMoved = false
                        rectSize = 45
                        paint.textSize = 30f
                        updateRect(false)
                    } else if (answer.contains(rectX, rectY)) {
                        paint.textSize = 30f
                        rectSize = 45
                        playSound(context, popID)
                        answer.add(this)
                        updateRect(false)
                    } else {
                        flinging = true
                    }
                }
                dragging = false
            }
            MotionEvent.ACTION_MOVE -> {
                if (dragging) {
                    rectX = mouseX
                    rectY = mouseY
                    updateRect(true)
                }
            }
        }
        return true
    }

    private fun updateRect(isOffset: Boolean) {
        val offset = if (isOffset) 40f else 0f
        rect = RectF(rectX - offset, rectY - offset, rectX + rectSize - offset, rectY + rectSize - offset)
    }

    private fun vibrate(duration: Long) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibratorManager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            vibratorManager.defaultVibrator.vibrate(VibrationEffect.createOneShot(duration, VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            @Suppress("DEPRECATION")
            val v = context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            v.vibrate(duration)
        }
    }

    fun anim() {
        if (flinging) {
            rectX += velX
            rectY += velY
            updateRect(true)
            if (transparency > 0) {
                transparency -= 5
            }
        } else if (movingLeft || movingRight) {
            rectX += velX
            if ((movingLeft && rectX <= targetX) || (movingRight && rectX >= targetX)) {
                rectX = targetX
                movingLeft = false
                movingRight = false
                velX = 0f
            }
            updateRect(false)
        }
    }

    fun getValue(): Int = value
    fun isSelected(): Boolean = dragging

    fun move(x: Float, y: Float) {
        targetX = x
        rectY = y
        if (rectX > x) {
            movingLeft = true
            movingRight = false
            velX = -speed
        } else {
            movingRight = true
            movingLeft = false
            velX = speed
        }
        updateRect(false)
    }
}
